Sources/Bot/Services/DispatchService.swift

```diff
@@ -116,18 +122,27 @@ public final class DispatchService {
             gitHubAPI: gitHubAPI,
             scheduler: scheduler
         )
+
+        // Forward events about creation and subsequent state changes of the new MergeService to our lifecycleObserver
         mergeServiceLifecycleObserver.send(value: .created(mergeService))
         mergeService.state.producer
+            .skipRepeats()
             .observe(on: scheduler)
-            .startWithValues { [weak self, service = mergeService, logger = logger] state in
+            .startWithValues { [weak self, service = mergeService] state in
                 self?.mergeServiceLifecycleObserver.send(value: .stateChanged(service))
-                if state.status == .idle {
-                    logger.log("👋 MergeService for target branch `\(targetBranch)` is idle, destroying")
-                    self?.mergeServices.modify { dict in
-                        dict[targetBranch] = nil
-                    }
-                    self?.mergeServiceLifecycleObserver.send(value: .destroyed(service))
+            }
+        // Observe idle states to clean up dormant MergeServices only after they have been idle for too long
+        mergeService.state.producer
+            .filter { $0.status == .idle }
+            .debounce(self.idleMergeServiceCleanupDelay, on: scheduler)
```

Comment:
🤔 just re-read my PR and realized, shouldn't those two lines be actually inverted? – If I `filter` first, there's not much to `debounce`, while what we want is to cancel the cleanup if the state goes out of `.idle` state

Reply:
if what we want is to wait until state is `idle` and then delay it then I think the order is correct.

Reply:
Assuming the cleanupDelay is 5mn, I think if the state machine goes into `.idle` at T=0 for 2 minutes but then receives new PRs in its queue at T=2mn and thus goes back to `.ready` -> `.integrating` and stays there for 20mn… Then we don't want to kill the `MergeService` at T=5mn while it's in the `.integrating` state just because it's 5mn after the last `.idle`.

What we want instead is to only kill the service it has remained in the `.idle` state for 5 minutes without going into a new state during those 5 minutes and thus been considered inactive for too long. Hence why I think the order of those lines might be wrong here

Reply:
> What we want instead is to only kill the service it has remained in the .idle state for 5 minutes without going into a new state during those 5 minutes

I'm not sure if that's how debounce works but if we can validate it with tests then it's ok

Reply:
FYI I wrote the test as you rightfully suggested (57e4501b58fa796aad0356223390506cd2553fba) and it indeed passes with `.debounce().filter()` but fail with `.filter().debounce()`, confirming my hypothesis 👍 

* The latter (`filter().debounce()`) indeed destroys the `MergeService` too soon especially if it was in `.idle` at some point and then it switches to `runningStatusChecks` but stays there too long (the `filter(.idle).debounce(T)` would then destroy the `MergeService` at that point while that's not what we want)
* While the former (`debounce(T).filter(.idle)`) and now current implementation properly destroys the service ONLY if it doesn't leave the `.idle` state _at all_ for the given delay – so exactly what we want to achieve 👍

Reply:
Not sure why `debounce` is needed instead of `delay`. 
I think the correct operator sequence is `filter().delay()`.

Reply:
@inamiy the goal is to cancel the destruction if we go out of idle state before the end of the delay. if we use `delay` it won't be cancelled will it? While my understanding of debounce is that it will cancel the destruction of the service if an event make use go out of the idle state before the delay, right?

Reply:
@AliSoftware 
> to avoid constantly killing and recreating them

I missed this PR body.
As you mentioned, it's probably better to add comment to disambiguate the use of `debounce` from `delay`.

Reply:
@AliSoftware  https://github.com/babylonhealth/Wall-E/pull/48#discussion_r358983938
I now also agree with `debounce(T).filter(.idle)` is the correct approach.